rpm-ostree-bisect

```diff
@@ -0,0 +1,347 @@
+        pull_commit_history(deployment, repo)
+
+        # initialize data
+        commits_info = initialize_commits_info(repo, 
+                                               badcommit, 
+                                               goodcommit)
+
+        # Write data to file
+        write_data(datafile, commits_info, testscript)
+
+        # write/enable systemd unit
+        with open(SYSTEMD_UNIT_FILE, 'w') as f:
+            f.write(SYSTEMD_UNIT_CONTENTS)
+        cmd = ['/usr/bin/systemctl', 'daemon-reload']
+        subprocess.run(cmd)
+        cmd = ['/usr/bin/systemctl', 'enable', SYSTEMD_UNIT_NAME]
+        subprocess.run(cmd)
+        return 0
+
+    else:
+        # load data
+        commits_info, testscript = load_data(datafile)
+
+        # run test
+        completed = subprocess.run(testscript, shell=True, check=False)
+        if completed.returncode == 0:
+            success = True
+        else:
+            success = False
+
+        # update and write data
+        commit = deployment.get_csum()
+        if success:
+            for c in reversed(commits_info.keys()):
+                commits_info[c]['status'] = 'GOOD' 
+                if c == commit:
+                    commits_info[c]['heuristic'] = 'TESTED' 
+                    break
+        else:
+            for c in commits_info.keys():
+                commits_info[c]['status'] = 'BAD' 
+                if c == commit:
+                    commits_info[c]['heuristic'] = 'TESTED' 
+                    break
+        write_data(datafile, commits_info, testscript)
+
+        # Find list of unknown status commits
+        unknowns = []
+        lastbad = None
+        firstgood = None
+        for commitid in commits_info.keys():
+            status = commits_info[commitid]['status'] 
+            if status == 'BAD':
+                lastbad = commitid
+            elif status == 'UNKNOWN':
+                unknowns.append(commitid)
+            elif status == 'GOOD':
+                firstgood = commitid
+                break
+
+        if len(unknowns) == 0:
+            # We're done!
+            log("Last known good commit:\n  %s : %s" %
+                  (firstgood, commits_info[firstgood]['version']))
+            log("First known bad commit:\n  %s : %s" %
+                  (lastbad, commits_info[lastbad]['version']))
+            # print out some db info
+            pull_commit_history(deployment, repo)
+            cmd = ['/usr/bin/rpm-ostree', 'db', 'diff',
+                   firstgood, lastbad]
+            subprocess.run(cmd, check=False)
+            return 0
+
+        # Bisect for new test commit id
+        newcommitid = unknowns[int(len(unknowns)/2)]
+
+        log("Deploying new commit for testing\n  %s : %s" % 
+              (newcommitid, commits_info[newcommitid]['version']))
+
+        cmd = ['/usr/bin/rpm-ostree', 'deploy', newcommitid]
+        log("Trying to run '%s'" % cmd)
+        tries = 30 # retry if download timesout
```

Comment:
Heh, 30 seems a bit high though... maybe e.g. 7? If it's failing after that, it seems more likely to not be a download issue, it feels?

Reply:
i was originally running this on F27 where we had the http2 issue. and since we're using ostree from the in-tree version we don't always have ostree with http2 disabled.